### PR TITLE
Fix exemplars being added to gauge metrics in the prometheus exporter

### PR DESCRIPTION
**Pull request by trthomps**
Prometheus Gauge metrics don't support exemplars and while `addGaugeMetric()` doesn't add them, `addSumMetric()` will if the metric is monotonic.  This causes the prometheus client to throw an error:

```
* error collecting metric Desc{fqName: "http_server_request_body_size_bytes", help: "Measures size of RPC request messages (uncompressed).", constLabels: {}, variableLabels: {net_protocol_name,net_protocol_version,http_method,http_route,http_scheme,net_host_name,net_host_port,otel_scope_name,otel_scope_version}}: cannot inject exemplar into Gauge, Summary or Untyped
```

### CODE REVIEW

**Comment by dashpole** [comment]
Would be good to have a test

**Comment by trthomps** [comment]
> Would be good to have a test

@dashpole Added a test, let me know if you know of a better way to test this, I did step through the debugger and exemplars are getting added to the `DataPoint`.

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 ### Fixed
 
 - Global MeterProvider registration unwraps global instrument Observers, the undocumented Unwrap() methods are now private. (#5881)
+- `go.opentelmetry.io/otel/exporters/prometheus` now doesn't try to add exemplars to Gauge metrics, which is unsupported. (#5912)
```
Comment (dmathieu) [inline CHANGELOG.md:24]:
```suggestion
- Fix `go.opentelemetry.io/otel/exporters/prometheus` trying to add exemplars to Gauge metrics, which is unsupported. (#5912)
```

File: exporters/prometheus/exporter.go [hunk 2]
```diff
@@ -277,7 +277,11 @@ func addSumMetric[N int64 | float64](ch chan<- prometheus.Metric, sum metricdata
 			otel.Handle(err)
 			continue
 		}
-		m = addExemplars(m, dp.Exemplars)
+		// GaugeValues don't support Exemplars at this time
+		// https://github.com/prometheus/client_golang/blob/aef8aedb4b6e1fb8ac1c90790645169125594096/prometheus/metric.go#L199
+		if valueType == prometheus.CounterValue {
```
Comment (dmathieu) [inline exporters/prometheus/exporter.go:282]: Would it be clearer and more aligned with the comment to do the opposite condition, to emphasize that we only want to skip exemplars for gauges?

```suggestion
		if valueType != prometheus.GaugeValue {
```

File: exporters/prometheus/exporter_test.go [hunk 3]
```diff
@@ -431,6 +432,47 @@ func TestPrometheusExporter(t *testing.T) {
...
+					sdktrace.WithSampler(sdktrace.AlwaysSample()),
+					sdktrace.WithResource(resource.NewWithAttributes(
+						semconv.SchemaURL,
+						semconv.ServiceNameKey.String("example"),
+					)),
+				)
+				tracer := tp.Tracer("example")
+				ctx, span := tracer.Start(ctx, "example")
+				defer span.End()
+
```
Comment (dmathieu) [inline exporters/prometheus/exporter_test.go:449]: Why do we need traces in a metrics exporter?

```suggestion
```

Comment (dashpole) [reply exporters/prometheus/exporter_test.go:449]: I assume you did this to generate a sampled trace context.  You can do that without using the trace SDK, similar to https://github.com/open-telemetry/opentelemetry-go/blob/7a153a01c6f75f05b63909bd08d823de65939bd2/exporters/prometheus/exporter_test.go#L976

File: exporters/prometheus/exporter_test.go [hunk 4]
```diff
@@ -431,6 +432,47 @@ func TestPrometheusExporter(t *testing.T) {
 				counter.Add(ctx, 5, otelmetric.WithAttributeSet(attrs2))
 			},
 		},
+		{
+			name:         "monotonic sum with exemplars",
```
Comment (dashpole) [inline exporters/prometheus/exporter_test.go:436]:
```suggestion
			name:         "non-monotonic sum does not add exemplars",
```

This is actually testing an updown (non-monotonic) counter, and we are checking that it does not add exemplars.  Consider renaming the test file as well.

File: exporters/prometheus/exporter_test.go (hunk 3 above)
Comment (trthomps) [reply exporters/prometheus/exporter_test.go]: Yes, exactly.  I found this bug using a middleware for gin that shares the context from gin for both traces and metrics reporting.  Agree it doesn't make sense normally attach trace information to a metric context.

**Comment by trthomps** [comment]
Thanks so much for working with me on this @dashpole.

File: exporters/prometheus/exporter_test.go [hunk 5]
```diff
@@ -431,6 +432,36 @@ func TestPrometheusExporter(t *testing.T) {
 				counter.Add(ctx, 5, otelmetric.WithAttributeSet(attrs2))
 			},
 		},
+		{
+			name:         "non-monotonic sum does not add exemplars",
+			expectedFile: "testdata/monotonic_sum_with_exemplars.txt",
+			recordMetrics: func(ctx context.Context, meter otelmetric.Meter) {
```
Comment (dashpole) [inline exporters/prometheus/exporter_test.go:438]: This test doesn't fail without your fix.  You should call Add with a sampled trace context, as described in https://github.com/open-telemetry/opentelemetry-go/pull/5912#discussion_r1812909317.  That does not rely on the trace SDK (just the API, which is OK)

File: exporters/prometheus/exporter_test.go [hunk 6]
```diff
@@ -23,6 +23,7 @@ import (
 	otelmetric "go.opentelemetry.io/otel/metric"
 	"go.opentelemetry.io/otel/sdk/metric"
 	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
```
Comment (dashpole) [inline exporters/prometheus/exporter_test.go:26]: remove this import now :)

File: exporters/prometheus/exporter_test.go [hunk 7]
```diff
@@ -431,6 +432,36 @@ func TestPrometheusExporter(t *testing.T) {
 				counter.Add(ctx, 5, otelmetric.WithAttributeSet(attrs2))
 			},
 		},
+		{
+			name:         "non-monotonic sum does not add exemplars",
+			expectedFile: "testdata/monotonic_sum_with_exemplars.txt",
```
Comment (dashpole) [inline exporters/prometheus/exporter_test.go:437]: please rename the file to match the test name

File: exporters/prometheus/exporter_test.go (hunk 5 above)
Comment (trthomps) [reply exporters/prometheus/exporter_test.go:437]: confirmed it fails without the fix now, thanks again for the help.

**Review by dashpole — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Pull request merged**